ufmt/core.py

```diff
@@ -2,118 +2,178 @@
 # Licensed under the MIT license
 
 import logging
-from dataclasses import dataclass
 from functools import partial
 from pathlib import Path
 from typing import List, Optional
+from warnings import warn
 
-from black import (
-    decode_bytes,
-    find_pyproject_toml,
-    format_file_contents,
-    Mode,
-    NothingChanged,
-    parse_pyproject_toml,
-    TargetVersion,
-)
+from black import format_file_contents, NothingChanged
 from moreorless.click import unified_diff
 from trailrunner import Trailrunner
-from usort.config import Config as UsortConfig
-from usort.sorting import usort_string
+from usort import usort
 
 from .config import ufmt_config
+from .types import (
+    BlackConfig,
+    BlackConfigFactory,
+    Encoding,
+    FileContent,
+    Result,
+    UsortConfig,
+    UsortConfigFactory,
+)
+from .util import make_black_config, read_file, write_file
 
 LOG = logging.getLogger(__name__)
 
 
-@dataclass
-class Result:
-    path: Path
-    changed: bool = False
-    written: bool = False
-    diff: Optional[str] = None
-
-
-def ufmt_string(
+def ufmt_bytes(
     path: Path,
-    content: str,
+    content: FileContent,
+    *,
+    black_config: BlackConfig,
     usort_config: UsortConfig,
-    black_config: Optional[Mode] = None,
-) -> str:
-    content = usort_string(content, usort_config, path)
-    mode = black_config or Mode()
+    encoding: Encoding = "utf-8",
+) -> FileContent:
+    """
+    Format arbitrary bytes for the given path.
+
+    Requires passing valid config objects for both black and µsort. If the given path
+    represents a type stub (has a ``.pyi`` suffix), the black config object will be
+    updated to set ``is_pyi = True``.
+
+    Note: content will be decoded before passing to black, and re-encoded to bytes
+    again. Be sure to pass a known-valid unicode encoding for the content.
+    :func:`ufmt.util.read_file` can be used to both read bytes from disk, and make a
+    best guess at file encodings.
+    """
+    result = usort(content, usort_config, path)
 
     if path.suffix == ".pyi":
-        mode.is_pyi = True
+        black_config.is_pyi = True
 
     try:
-        content = format_file_contents(content, fast=False, mode=mode)
-    except NothingChanged:
-        pass
-
-    return content
-
-
-def make_black_config(path: Path) -> Mode:
-    config_file = find_pyproject_toml((str(path),))
-    if not config_file:
-        return Mode()
-
-    config = parse_pyproject_toml(config_file)
+        content_str = result.output.decode(encoding)
+        content_str = format_file_contents(content_str, fast=False, mode=black_config)
+        content = content_str.encode(encoding)
```

Comment:
It looks like black has crlf detection as well, which is being bypassed here.

Reply:
I was a bit unsure about this. In black's `format_file_in_place`, it calls `decode_bytes` which returns the newline value, but as far as I can tell, black doesn't do anything with this in `format_file_contents`. 

Though on further digging, it seems that black is depending on implicit conversion of newlines in both the read and write path...

Reply:
That's my read too; black _will_ normalize the newlines based on whatever the end of the first line is.  I think usort will leave a mixed file alone, except for imports which get the detected newline style (from the first line, I'd assume).

1. It appears that black documents that it expects to only see LF: https://github.com/psf/black/blob/main/src/black/__init__.py#L1205
2. Do we like the normalization behavior?  (If we do, I'd like usort to do the same and have the Result also pass along the libcst-detected newline, and maybe do the normalization as well -- sounds like a bugfix?)